File: 코딩헬스장/아스키코드/아스키코드.py
# 문자를 입력받으면 아스키코드 10진수 int로 전환하고 0~127 사이의 int 값을 입력받으면 아스키코드에 해당하는 문자로 전환하는 함수를 작성하세요.
# TypeError: ord() expected a character, but string of length 2 found -> len(i) == 1

def ascii_code(i):
    if isinstance(i, str) and len(i) == 1:
        return ord(i)
    elif isinstance(i, int) and 0 <= i <= 127:
        return chr(i)
    else:
        return "error: invalid input."

# Test
import unittest

class TestAsciiCode(unittest.TestCase):
    def test_char_to_ascii(self):
        self.assertEqual(ascii_code('A'), 65)
        self.assertEqual(ascii_code('a'), 97)
        self.assertEqual(ascii_code('4'), 52)

    def test_ascii_to_char(self):
        self.assertEqual(ascii_code(65), 'A')
        self.assertEqual(ascii_code(97), 'a')
        self.assertEqual(ascii_code(48), '0')

    def test_invalid_input(self):
        self.assertEqual(ascii_code('AB'), "error: invalid input.")
        self.assertEqual(ascii_code(128), "error: invalid input.")
        self.assertEqual(ascii_code(-1), "error: invalid input.")

if __name__ == '__main__':
    unittest.main()

File: 코딩헬스장/아스키코드/아스키코드.kt
fun asciiCode(input: Any): Any {
    return when (input) {
        is Char -> input.toInt()
        is Int -> {
            if (input in 0..127) input.toChar() else "Invalid input"
        }
        else -> "Invalid input"
    }
}

fun main() {
    println(asciiCode('A') == 65)
    println(asciiCode('a') == 97)
    println(asciiCode('0') == 48)
    println(asciiCode('4') == 52)
}

// is 키워드 : input 타입 검사하고 해당 타입으로 스마트 캐스팅
// 범위 연산자 in : input이 0~127 사이에 있는지 검사
// when 키워드 : Java의 switch 문보다 더 강력하고 유연
// 타입 안정성 : input이 Char, Int 타입이 아닌 경우 "Invalid input" 반환